### PR TITLE
Query Exceptionless Fix

### PR DESCRIPTION
**Pull request by bchong95**
# Query Exceptionless Fix

### Bug

Queries were returning partial / inconsistent results when encountering 429s. The reason for this is that we were buffering exceptions and not setting the state correctly which ended up nulling out the continuation token. This ended up making the query iterator think that there was no more results instead of executing another roundtrip and getting back the 429. 

### Fix

For parallel (non order by) the fix was to just refactor the way we enumerate documents. The problem with parallel is that abstracted pages from the query pipeline which meant that parallel would always receive the next page of results  when it got the end of one page. If that next page was a 429 we would buffer the 429 and send the results we have so far (but with a null continuation token). The solution was to return a continuation token at the end of each page and not buffer the 429. When the user comes back for the next page they will get the 429. ORDER BY was a trickier, since you will need to grab multiple pages in order to serve the sort order. For this I bit the bullet and returned the 429 and threw away any partial work. In the future we can add more buffering mechanisms to try and return the partial results.

### Tests

None of the existing tests caught this situation. This is mainly because there are so many combinations that are hard to generate through unit tests. The fix was to wire through some test settings that will let the individual item producers know if we want to simulate edge cases like 429s and empty pages. From there the integration test hit all the failure cases.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs [hunk 1]
```diff
@@ -107,34 +114,46 @@ protected override string ContinuationToken
...
+
+                    continuationToken = JsonConvert.SerializeObject(orderByContinuationTokens, new JsonSerializerSettings()
                     {
                         StringEscapeHandling = StringEscapeHandling.EscapeNonAscii,
-                    }) : null;
+                    });
+
+                    // Newtonsoft has a bug where non ascii characters aren't escaped for custom POGOs
+                    // so the workaround is to double serialize
+                    continuationToken = JsonConvert.SerializeObject(JToken.Parse(continuationToken), new JsonSerializerSettings()
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: >continuationToken [](start = 20, length = 17)

add a test for non ascii continuation tokens.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CrossPartitionQueryTests.cs [hunk 2]
```diff
@@ -1346,6 +1381,94 @@ public async Task TestBasicCrossPartitionQuery()
...
+        public async Task TestEmptyPages()
+        {
+            int seed = (int)(DateTime.UtcNow - new DateTime(1970, 1, 1)).TotalSeconds;
+            uint numberOfDocuments = 100;
+            QueryOracle.QueryOracleUtil util = new QueryOracle.QueryOracle2(seed);
+            IEnumerable<string> documents = util.GetDocuments(numberOfDocuments);
+
+            await this.CreateIngestQueryDelete(
+                ConnectionModes.Direct,
+                CollectionTypes.SinglePartition /*| CollectionTypes.MultiPartition*/,
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CrossPartitionQueryTests.cs]: >/*| CollectionTypes.MultiPartition*/, [](start = 48, length = 37)

uncomment.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Query/ItemProducerTreeUnitTests.cs [hunk 3]
```diff
@@ -1,287 +1,290 @@
...
-            Assert.IsFalse(itemProducerTree.HasMoreResults);
-        }
-    }
-}
\ No newline at end of file
+﻿////------------------------------------------------------------
+//// Copyright (c) Microsoft Corporation.  All rights reserved.
+////------------------------------------------------------------
+
+//namespace Microsoft.Azure.Cosmos.Tests
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Query/ItemProducerTreeUnitTests.cs]: Add these back in

File: Microsoft.Azure.Cosmos/src/Query/Core/ItemProducerTree/ItemProducerTree.cs [hunk 4]
```diff
@@ -656,7 +612,25 @@ private async Task<(bool successfullyMovedNext, QueryResponseCore? failureRespon
...
+                                if (!succeeded)
+                                {
+                                    throw new InvalidOperationException("Unable to get the first page for the child document producers");
+                                }
+
+                                if (replacementItemProducerTree.ItemsLeftInCurrentPage != 0)
+                                {
+                                    // Order By requires at least one item in the tree so that the comparator will work.
+                                    break;
+                                }
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/Core/ItemProducerTree/ItemProducerTree.cs]: Make sure there is enough split test coverage.

File: Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.Query.Core
+{
+    internal sealed class TestSettings
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs]: 1. Passing a test setting object into product code is usually considered bad design. 
2. This doesn't need to exist. You can accomplish the same thing with mocking or a custom RequestHandler.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosParallelItemQueryExecutionContext.cs [hunk 6]
```diff
@@ -47,19 +48,22 @@ internal sealed class CosmosParallelItemQueryExecutionContext : CosmosCrossParti
         /// <param name="maxConcurrency">The max concurrency</param>
         /// <param name="maxBufferedItemCount">The max buffered item count</param>
         /// <param name="maxItemCount">Max item count</param>
+        /// <param name="testSettings">Test settings.</param>
         private CosmosParallelItemQueryExecutionContext(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosParallelItemQueryExecutionContext.cs]: Please update changelog

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 7]
```diff
@@ -452,31 +369,48 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
...
+
+                        if (failureResponse.HasValue)
+                        {
+                            return TryCatch<bool>.FromException(
+                                new CosmosException(
+                                    statusCode: failureResponse.Value.StatusCode,
+                                    subStatusCode: (int)failureResponse.Value.SubStatusCode.GetValueOrDefault(0),
+                                    message: failureResponse.Value.ErrorMessage,
+                                    activityId: failureResponse.Value.ActivityId,
+                                    requestCharge: 0));
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: Failed responses can still have a request charge.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 8]
```diff
@@ -452,31 +369,48 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
...
+                        }
+
+                        if (!movedToNextPage)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: Shouldn't this be combined with the above exception? Is there a scenario where moved next page failed and there is not an exception?

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 9]
```diff
@@ -452,31 +369,48 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
...
-                if (itemProducerTree.HasMoreResults)
-                {
-                    this.itemProducerForest.Enqueue(itemProducerTree);
-                }
+                this.itemProducerForest.Enqueue(itemProducerTree);
+            }
+
+            if (this.itemProducerForest.Count == 0)
+            {
+                return TryCatch<bool>.FromException(new Exception("Failed to initialize cross partition query execution context."));
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: Should this be a CosmosException? Should we add more info the the exception like forest count before and after the filter?

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs [hunk 10]
```diff
@@ -107,34 +114,46 @@ protected override string ContinuationToken
...
+                    {
+                        StringEscapeHandling = StringEscapeHandling.EscapeNonAscii,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: Make the JsonSerializerSettings a static readonly.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs [hunk 11]
```diff
@@ -107,34 +114,46 @@ protected override string ContinuationToken
...
+                            },
+                            orderByQueryResult.OrderByItems,
+                            orderByQueryResult.Rid,
+                            this.ShouldIncrementSkipCount(itemProducer) ? this.skipCount + 1 : 0,
+                            filter);
+
+                        return orderByContinuationToken;
+                    });
+
+                    continuationToken = JsonConvert.SerializeObject(orderByContinuationTokens, new JsonSerializerSettings()
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: Does the JsonSerializer also have the bug that requires the double serialize?

File: Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public TestSettings(bool simulate429s, bool simulateEmptyPages)
+        {
+            this.Simulate429s = simulate429s;
+            this.SimulateEmptyPages = simulateEmptyPages;
+        }
+
+        public bool Simulate429s { get; }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs]: >Simulate429s [](start = 20, length = 12)

429 to throttles

File: Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs (hunk 5 above)
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs]: >TestSettings [](start = 26, length = 12)

TestInjection

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 13]
```diff
@@ -70,6 +68,8 @@ internal abstract class CosmosCrossPartitionQueryExecutionContext : CosmosQueryE
         /// </summary>
         private readonly long actualMaxBufferedItemCount;
 
+        private readonly TestSettings testSettings;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: >private readonly TestSettings testSettings; [](start = 8, length = 43)

add a comment

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 14]
```diff
@@ -212,7 +205,7 @@ internal abstract class CosmosCrossPartitionQueryExecutionContext : CosmosQueryE
         /// <summary>
         /// Gets a value indicating whether the context still has more results.
         /// </summary>
-        private bool HasMoreResults => this.FailureResponse != null || (this.itemProducerForest.Count != 0 && this.CurrentItemProducerTree().HasMoreResults);
+        private bool HasMoreResults => this.itemProducerForest.Count != 0 && this.CurrentItemProducerTree().HasMoreResults;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: >this.itemProducerForest.Count != 0 [](start = 39, length = 34)

()

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs (hunk 11 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: Probably 

---
In reply to: [346820683](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/999#discussion_r346820683) [](ancestors = 346820683)

File: Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs (hunk 5 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/TestSettings.cs]: In a future iteration we can clean it up. We could even just have a composition of an existing mock to avoid creating a new mock. 

---
In reply to: [346813511](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/999#discussion_r346813511) [](ancestors = 346813511)

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs (hunk 8 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: Yeah if you call move next at the end of the query. 

---
In reply to: [346816892](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/999#discussion_r346816892) [](ancestors = 346816892)

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosParallelItemQueryExecutionContext.cs [hunk 15]
```diff
@@ -48,19 +48,22 @@ internal sealed class CosmosParallelItemQueryExecutionContext : CosmosCrossParti
...
             CosmosQueryContext queryContext,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosParallelItemQueryExecutionContext.cs]: Please update changelog

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosParallelItemQueryExecutionContext.cs]: Done

**Comment by kirankumarkolli** [comment]
#1168
